src/ram_key_value_store/mod.rs

```diff
@@ -0,0 +1,103 @@
+use super::key_value_store::Key;
+use super::key_value_store::KeyValueStore;
+use super::key_value_store::Size;
+use core::cell::Cell;
+use core::mem;
+use core::mem::MaybeUninit;
+
+#[cfg(test)]
+mod test;
+
+pub struct RamKeyValueStoreElement {
+    pub key: Key,
+    pub size: Size,
+}
+
+impl RamKeyValueStoreElement {
+    pub fn new<T: Sized>(key: Key) -> Self {
+        Self {
+            key: key,
+            size: mem::size_of::<T>() as u8,
+        }
+    }
+}
+
+pub struct RamKeyValueStore<'a> {
+    ram: &'a mut [Cell<u8>],
+    elements: &'a [RamKeyValueStoreElement],
+}
+
+impl<'a> RamKeyValueStore<'a> {
+    pub fn new(ram: &'a mut [Cell<u8>], elements: &'a [RamKeyValueStoreElement]) -> Self {
+        let required_size: usize = elements.iter().map(|x| x.size as usize).sum();
+        assert!(ram.len() == required_size, "RAM size doesn't match");
+
+        Self { ram, elements }
+    }
+}
+
+impl<'a> KeyValueStore for RamKeyValueStore<'a> {
+    fn read<T: crate::key_value_store::SafelyDeserializable + Sized>(&self, key: Key) -> T {
+        let mut offset = 0 as usize;
+
+        for element in self.elements.iter() {
+            if key == element.key {
+                assert!(mem::size_of::<T>() == element.size as usize, "Invalid size");
+                assert!(
+                    <T>::can_deserialize_from(&self.ram[offset..]),
+                    "Invalid contents"
+                );
+
+                unsafe {
+                    let mut value = MaybeUninit::<T>::zeroed();
+
+                    for i in 0..element.size {
+                        let byte = self.ram[offset + i as usize].get();
+                        let raw_pointer = value.as_mut_ptr() as *mut u8;
+                        *raw_pointer.offset(i as isize) = byte;
+                    }
+
+                    return value.assume_init();
```

Comment:
```suggestion
                let mut value = MaybeUninit::<T>::zeroed();

                let dst_raw_pointer = value.as_mut_ptr() as *mut u8;
                let src_raw_pointer = self.ram[offset..].as_ptr() as *const u8;

                unsafe {
                    core::ptr::copy_nonoverlapping(src_raw_pointer, dst_raw_pointer, element.size as usize);

                    return value.assume_init();
                };
```
No for loop, but idk if it is really that much cleaner, I guess we could rename `core::ptr::copy_nonoverlapping` to `memcpy` as that is what it is, and that could help?

Safety:
I'm assuming that the tests prove that `self.ram[offset..].as_ptr()` works as we would expect it to (move the ptr by offset). As for the `as_ptr() as *const u8` this should be fine as both `Cell` and `UnsafeCell` are `#[repr(transparent)]`

Reply:
Hmm, I'll play with that. I wasn't sure if I could use any of the `memcpy`ish functions since each byte is technically a `Cell`, but it does make sense that we should be able to just ignore that for this since the underlying type is `u8`.

Reply:
I think we get to ignore the `Cell` because it is `repr(transparent)`